### PR TITLE
Instrumenation.AWSLambda: Drop netstandard2, lose Newtonsoft (non-)dependency

### PR DESCRIPTION
**Pull request by Oberon00**
Closes #1273 (alternative to that PR)

Related to a similar change in the XRay dependency: #1092

## Changes

Quoting the CHANGELOG:

* The minimum required .NET version is now .NET Core 3.1 (previously any netstandard2.0-compliant).
  As the minimum .NET version in AWS Lambda is already 6.0, no compatibility problems are expected.

  This resolves a warning that some dependency analyzers may produce where this package would transitively
  depend on a vulnerable version of `Newtonsoft.Json` through
  [Amazon.Lambda.APIGatewayEvents](https://www.nuget.org/packages/Amazon.Lambda.APIGatewayEvents/2.4.1#dependencies-body-tab).

  This also avoids a potential issue where the instrumentation would try to call a Newtonsoft.Json function
  when no other package nor the app itself depends on Newtonsoft.Json, since the transitive dependency would be
  ignored unless using application were compiled against a TargetFramework older than Core 3.1.

For significant contributions please make sure you have completed the following items:

* [X] Appropriate `CHANGELOG.md` updated for non-trivial changes
* ~~[ ] Design discussion issue #~~ N/A
* ~~[ ] Changes in public API reviewed~~ API remains identical

### CODE REVIEW

**Comment by Oberon00** [comment]
BTW, it seems we are using an outdated tool dotnet-format instead of the built-in dotnet format subcommand. The former seems to be completely confused by raw string literals.

**Comment by Oberon00** [comment]
One unrelated build failure:

```
2023-07-18T15:36:33.5008155Z [xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.5+1caef2f33e (64-bit .NET 7.0.9)
2023-07-18T15:36:34.0200913Z [xUnit.net 00:00:00.53]   Discovering: OpenTelemetry.Instrumentation.Hangfire.Tests
2023-07-18T15:36:34.0589441Z [xUnit.net 00:00:00.57]   Discovered:  OpenTelemetry.Instrumentation.Hangfire.Tests
2023-07-18T15:36:34.0624229Z [xUnit.net 00:00:00.57]   Starting:    OpenTelemetry.Instrumentation.Hangfire.Tests
2023-07-18T15:36:34.5551443Z [xUnit.net 00:00:01.05]     OpenTelemetry.Instrumentation.Hangfire.Tests.HangfireInstrumentationJobFilterAttributeTests.Should_Create_Activity_With_Status_Error_When_Job_Failed [FAIL]
2023-07-18T15:36:34.5552699Z [xUnit.net 00:00:01.05]       System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
2023-07-18T15:36:34.5554807Z [xUnit.net 00:00:01.05]       Stack Trace:
2023-07-18T15:36:34.5555773Z [xUnit.net 00:00:01.05]            at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
2023-07-18T15:36:34.5556693Z [xUnit.net 00:00:01.05]            at System.Linq.Enumerable.All[TSource](IEnumerable`1 source, Func`2 predicate)
2023-07-18T15:36:34.5559627Z [xUnit.net 00:00:01.05]         D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\test\OpenTelemetry.Instrumentation.Hangfire.Tests\HangfireInstrumentationJobFilterAttributeTests.cs(157,0): at OpenTelemetry.Instrumentation.Hangfire.Tests.HangfireInstrumentationJobFilterAttributeTests.WaitJobProcessedAsync(String jobId, Int32 timeToWaitInSeconds)
2023-07-18T15:36:34.5563935Z [xUnit.net 00:00:01.05]         D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\test\OpenTelemetry.Instrumentation.Hangfire.Tests\HangfireInstrumentationJobFilterAttributeTests.cs(71,0): at OpenTelemetry.Instrumentation.Hangfire.Tests.HangfireInstrumentationJobFilterAttributeTests.Should_Create_Activity_With_Status_Error_When_Job_Failed()
2023-07-18T15:36:34.5566329Z [xUnit.net 00:00:01.05]         --- End of stack trace from previous location ---
2023-07-18T15:36:34.9320774Z   Failed OpenTelemetry.Instrumentation.Hangfire.Tests.HangfireInstrumentationJobFilterAttributeTests.Should_Create_Activity_With_Status_Error_When_Job_Failed [303 ms]
2023-07-18T15:36:34.9322116Z   Error Message:
2023-07-18T15:36:34.9322588Z    System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
2023-07-18T15:36:34.9322995Z   Stack Trace:
2023-07-18T15:36:34.9323391Z      at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
2023-07-18T15:36:34.9323923Z    at System.Linq.Enumerable.All[TSource](IEnumerable`1 source, Func`2 predicate)
2023-07-18T15:36:34.9326378Z    at OpenTelemetry.Instrumentation.Hangfire.Tests.HangfireInstrumentationJobFilterAttributeTests.WaitJobProcessedAsync(String jobId, Int32 timeToWaitInSeconds) in D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\test\OpenTelemetry.Instrumentation.Hangfire.Tests\HangfireInstrumentationJobFilterAttributeTests.cs:line 157
2023-07-18T15:36:34.9329362Z    at OpenTelemetry.Instrumentation.Hangfire.Tests.HangfireInstrumentationJobFilterAttributeTests.Should_Create_Activity_With_Status_Error_When_Job_Failed() in D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\test\OpenTelemetry.Instrumentation.Hangfire.Tests\HangfireInstrumentationJobFilterAttributeTests.cs:line 71
2023-07-18T15:36:34.9330839Z --- End of stack trace from previous location ---
2023-07-18T15:36:38.1045980Z [xUnit.net 00:00:04.61]   Finished:    OpenTelemetry.Instrumentation.Hangfire.Tests
```

File: src/OpenTelemetry.Instrumentation.AWSLambda/OpenTelemetry.Instrumentation.AWSLambda.csproj [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.AWSLambda/OpenTelemetry.Instrumentation.AWSLambda.csproj:4]: Can we change it to net6.0? netcoreapp3.1 is deprecated and might leads to issues in CI.

Comment (Oberon00) [reply src/OpenTelemetry.Instrumentation.AWSLambda/OpenTelemetry.Instrumentation.AWSLambda.csproj:4]: It already passed CI though (except for a timeout in the coverage build).

I wouldn't want to switch right to 6. If netcoreapp3.1 is a problem, I'd stay on netstandard2.0 and add a conditional dependency on System.Text.Json just like in https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1092

**Review by rypdal — APPROVED** [review]

File: src/OpenTelemetry.Instrumentation.AWSLambda/OpenTelemetry.Instrumentation.AWSLambda.csproj (hunk 1 above)
Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AWSLambda/OpenTelemetry.Instrumentation.AWSLambda.csproj:4]: > As the minimum .NET version in AWS Lambda is already 6.0, no compatibility problems are expected.

@Oberon00 Why do you not want to target net6.0, when the minimum runtime version required by ASWLambda is net6.0?

Comment (Oberon00) [reply src/OpenTelemetry.Instrumentation.AWSLambda/OpenTelemetry.Instrumentation.AWSLambda.csproj:4]: .NET 6 is not even 2 years old. And while the .NET 6 runtime is the oldest one supported by AWS, you should still be able compile your app against older versions and run them. Also, you could use a containerized runtime with an older version.

That's why I'd prefer not to use such a recent version as minimum, especially when it's only a bugfix.

Comment (Oberon00) [reply src/OpenTelemetry.Instrumentation.AWSLambda/OpenTelemetry.Instrumentation.AWSLambda.csproj:4]: I created PR #1273 as an alternative to this, where I do not touch the .NET version at all, staying on netstandard2.0, adding an explicit dependency on Newtonsoft.Json instead.

Probably the .NET version upgrade should be done after the next release of the package.

Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AWSLambda/OpenTelemetry.Instrumentation.AWSLambda.csproj:4]: .NET 6 is marked for LTS (long-term support) by .NET team and if AWS Lambda has chosen it as the minimum version, then it makes sense for its corresponding instrumentation library to do the same.

Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AWSLambda/OpenTelemetry.Instrumentation.AWSLambda.csproj:4]: For AWS Lambda, is it possible for a consuming application to be targeting something lower than .NET 6? If not, then why would you be bothered by any of the lower and older targets?

Comment (Oberon00) [reply src/OpenTelemetry.Instrumentation.AWSLambda/OpenTelemetry.Instrumentation.AWSLambda.csproj:4]: @utpilla https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1270#discussion_r1268962025:

> you should still be able compile your app against older versions and run them. Also, you could use a containerized runtime with an older version.

To elaborate on the containerized runtime: https://docs.aws.amazon.com/lambda/latest/dg/images-create.html

I do think we should raise the requirement to net6 eventually, but not for a bugfix.

Comment (Kielek) [reply src/OpenTelemetry.Instrumentation.AWSLambda/OpenTelemetry.Instrumentation.AWSLambda.csproj:4]: Lets do opposite.
1. Create PR with .NET 6 bump.
2. As a follow up, fix this issue.

Comment (Oberon00) [reply src/OpenTelemetry.Instrumentation.AWSLambda/OpenTelemetry.Instrumentation.AWSLambda.csproj:4]: If the upgrade to net6 is urgent, I'd prefer:

1. Merge a PR without a bump to net 6 (this or https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1273)
2. Publish a release
3. .NET 6 bump

**Pull request closed without merging**